File: Hdl21SchematicImporter/hdl21schematicimporter/circuit.py
""" 
# Circuit

Circuits data model extracted from schematics. 

Circuit data is often parsed from JSON-encoded text embedded in an SVG schematic. 
Thus it is crucial that the data model here matches that of the editor tool.
"""

from typing import List
from enum import Enum
from pydantic.dataclasses import dataclass


class PortDir(Enum):
    """# Signal/ Port Direction
    Including the `INTERNAL` variant for internal Signals"""

    INTERNAL = "INTERNAL"
    INPUT = "INPUT"
    OUTPUT = "OUTPUT"
    INOUT = "INOUT"


@dataclass
class Signal:
    """# Circuit Signal"""

    name: str
    portdir: PortDir


@dataclass
class Connection:
    """# Instance Connection
    A (port, signal) pair, specified in string names."""

    portname: str
    signame: str


@dataclass
class Instance:
    """# Circuit Instance"""

    name: str  # Instance Name
    of: str  # Instance-Of Code-String
    conns: List[Connection]  # Connections


@dataclass
class Circuit:
    """# Circuit
    The circuit-level content of a Schematic. This might alternatively be called a "Module". 
    Consists of collections of Signals, Instances of circuit elements, and connections there-between."""

    name: str  # Circuit Name
    prelude: str  # Code Prelude
    signals: List[Signal]  # Signals
    instances: List[Instance]  # Instances

File: Hdl21SchematicEditor/packages/EditorCore/src/circuit/circuit.ts
//
// # Circuit
//
// Circuits data model extracted from schematics.
//
// Circuit data is often stored as JSON-encoded text embedded in an SVG schematic.
// Thus it is crucial that the data model here matches that of any paired importer.
//

// # Signal/ Port Direction
// Including the `INTERNAL` variant for internal Signals
export enum PortDir {
  INTERNAL = "INTERNAL",
  INPUT = "INPUT",
  OUTPUT = "OUTPUT",
  INOUT = "INOUT",
}

// # Circuit Signal
export interface Signal {
  name: string;
  portdir: PortDir;
}

// # Instance Connection
// A (port, signal) pair, specified in string names.
export interface Connection {
  portname: string; // Port Name
  signame: string; // Connected Signal Name
}

// # Circuit Instance
export interface Instance {
  name: string; // Instance Name
  of: string; // Instance-Of Code-String
  conns: Array<Connection>; // Connections
}

// # Circuit
// The circuit-level content of a Schematic. This might alternatively be called a "Module".
// Consists of collections of Signals, Instances of circuit elements, and connections there-between.
export interface Circuit {
  name: string; // Circuit Name
  prelude: string; // Code Prelude
  signals: Array<Signal>; // Signals List
  instances: Array<Instance>; // Instances List
}
